base/shell/explorer/trayntfy.cpp

```diff
@@ -308,17 +313,226 @@ class CIconWatcher
     }
 };
 
+class CBalloonQueue
+{
+public:
+    static const int TimerInterval = 2000;
+    static const int BalloonsTimerId = 1;
+    static const int MinTimeout = 10000;
+    static const int MaxTimeout = 30000;
+    static const int CooldownBetweenBalloons = 2000;
+
+private:
+    struct Info
+    {
+        InternalIconData * pSource;
+        WCHAR szInfo[256];
+        WCHAR szInfoTitle[64];
+        WPARAM uIcon;
+        UINT uTimeout;
+
+        Info(InternalIconData * source)
+        {
+            pSource = source;
+            StrNCpy(szInfo, source->szInfo, _countof(szInfo));
+            StrNCpy(szInfoTitle, source->szInfoTitle, _countof(szInfoTitle));
+            uIcon = source->dwInfoFlags & 3;
+            if (source->dwInfoFlags & NIIF_USER)
+                uIcon = reinterpret_cast<WPARAM>(source->hIcon);
+            uTimeout = source->uTimeout;
+        }
+    };
+
+    HWND m_hwndParent;
+
+    CTooltips * m_tooltips;
+
+    CAtlList<Info> m_queue;
+
+    CToolbar<InternalIconData> * m_toolbar;
+
+    InternalIconData * m_current;
+    bool m_closed;
```

Comment:
Is it grouped intentionally?

Reply:
Yes. I leave m_current assigned up until the next icon shows or the cooldown elapses, so that other balloons being added in that space of time won't cause it to trigger early.

Reply:
By which I meant: m_closed is needed to distinguish between "a balloon is open" and "a balloon has been closed and we are in a short cooldown between balloons", while still being able to tell which icon showed the balloon so we can short-circuit and open a new balloon.

F*k writing this made me realize I introduced a bug in the cleanup.

Reply:
Hehe, try adding more verbose name :yum: